apps/api/src/notifications/notifications.service.ts

```diff
@@ -0,0 +1,314 @@
+    private readonly sendEmail: EmailService,
+    private readonly campaignService: CampaignService,
+    private readonly marketingNotificationsProvider: NotificationsProviderInterface<SendGridParams>,
+  ) {}
+
+  async sendConfirmConsentEmail(data: SendConfirmationDto) {
+    // Check if user is registered
+    let registered: Person | null
+    try {
+      registered = await this.personService.findByEmail(data.email)
+    } catch (e) {
+      return new BadRequestException('Failed to get email data')
+    }
+    // If registered and has given consent -> user is already subscribed
+    if (registered?.newsletter) return { message: 'Subscribed' }
+
+    // Check if un-registered consent already exists for this email
+    let unregisteredConsent: UnregisteredNotificationConsent | null
+    try {
+      unregisteredConsent = await this.prisma.unregisteredNotificationConsent.findFirst({
+        where: { email: data.email, consent: true },
+      })
+    } catch (e) {
+      return new BadRequestException('Failed to get email data')
+    }
+
+    // If unregistered consent exists -> user is already subscribed
+    if (unregisteredConsent?.consent) return { message: 'Subscribed' }
+
+    // Send confirmation email
+    try {
+      const sent = await this.prisma.emailSentRegistry.findFirst({
+        where: { email: data.email, type: EmailType.confirmConsent },
+      })
+      // Check if email was sent already in the past 1 minutes
+      const wasSent = this.emailWasSentRecently(sent, 1)
+
+      if (wasSent) return { message: 'Email Sent' }
+
+      // Secret hash to verify data in the subscribe route
+      const secret = this.generateHash()
+
+      // Save hash to record
+      registered
+        ? // Update
+          await this.prisma.person.update({
+            where: { id: registered.id },
+            data: { mailHash: secret },
+          })
+        : // Create/Update
+          await this.prisma.unregisteredNotificationConsent.upsert({
+            where: { email: data.email },
+            create: { mailHash: secret, email: data.email },
+            update: { mailHash: secret },
+          })
+
+      // Public subscribe link
+      const link =
+        this.config.get<string>('APP_URL', '') +
+        `/notifications/subscribe?hash=${secret}&email=${data.email}&consent=true`
+
+      // Prepare Email data
+      const recepient = { to: [data.email] }
+      const mail = new ConfirmConsentEmailDto({
+        subscribeLink: link,
+      })
+
+      await this.sendEmail.sendFromTemplate(mail, recepient)
+
+      sent
+        ? // Update new send date
+          await this.prisma.emailSentRegistry.update({
+            where: { id: sent.id },
+            data: { dateSent: new Date() },
+          })
+        : // Create new record for the email sent
+          await this.prisma.emailSentRegistry.create({
+            data: { email: data.email, type: EmailType.confirmConsent, dateSent: new Date() },
```

Comment:
feel free to use this.prisma.emailSentRegistry.upsert for less code

Reply:
I've tried. The problem is that upsert requires that the `{where}` clause includes only unique values. Email is not unique in the EmailSentRegistry Table, because we might want to keep track of different types of emails sent to the same email. This is why i used this approach